binderhub/build_local.py

```diff
@@ -0,0 +1,221 @@
+            If a build with the same name is already running when submitted,
+            only the last `log_tail_lines` number of lines will be fetched and
+            displayed to the end user. If not, all log lines will be streamed.
+        sticky_builds : ignored
+        """
+        self.q = q
+        self.repo_url = repo_url
+        self.ref = ref
+        self.name = name
+        self.image_name = image_name
+        self.push_secret = push_secret
+        self.main_loop = IOLoop.current()
+        self.memory_limit = memory_limit
+        self.memory_request = memory_request
+        self.appendix = appendix
+        self.log_tail_lines = log_tail_lines
+
+        self.git_credentials = git_credentials
+
+    @classmethod
+    def cleanup_builds(cls, kube, namespace, max_age):
+        app_log.debug("Not implemented")
+
+    def progress(self, kind: ProgressEvent.Kind, payload: str):
+        """
+        Put current progress info into the queue on the main thread
+        """
+        self.main_loop.add_callback(self.q.put, ProgressEvent(kind, payload))
+
+    def get_affinity(self):
+        raise NotImplementedError()
+
+    def submit(self):
+        """
+        Run a build to create the image for the repository.
+
+        Progress of the build can be monitored by listening for items in
+        the Queue passed to the constructor as `q`.
+        """
+        env = os.environ.copy()
+        if self.git_credentials:
+            env['GIT_CREDENTIAL_ENV'] = self.git_credentials
+
+        cmd = self.get_cmd()
+        app_log.info("Starting build: %s", " ".join(cmd))
+
+        try:
+            self.progress(ProgressEvent.Kind.BUILD_STATUS_CHANGE, ProgressEvent.BuildStatus.RUNNING)
+            for line in _execute_cmd(cmd, capture=True, env=env):
+                self._handle_log(line)
+            self.progress(ProgressEvent.Kind.BUILD_STATUS_CHANGE, ProgressEvent.BuildStatus.COMPLETED)
+        except subprocess.CalledProcessError as e:
+            self.progress(ProgressEvent.Kind.BUILD_STATUS_CHANGE, ProgressEvent.BuildStatus.FAILED)
+        except Exception as e:
+            app_log.exception("Error in watch stream for %s", self.name)
+            raise
+
+    def _handle_log(self, line):
+        try:
+            json.loads(line)
+        except ValueError:
+            # log event wasn't JSON.
+            # use the line itself as the message with unknown phase.
+            # We don't know what the right phase is, use 'unknown'.
+            # If it was a fatal error, presumably a 'failure'
+            # message will arrive shortly.
+            app_log.error("log event not json: %r", line)
+            line = json.dumps({
+                'phase': 'unknown',
+                'message': line,
+            })
+        self.progress(ProgressEvent.Kind.LOG_MESSAGE, line)
+
+    def stream_logs(self):
+        pass
+
+    def cleanup(self):
+        raise NotImplementedError()
+
+    def stop(self):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError()
```

`stop` should send appropriate signals to the repo2docker process so it stops, and this we should implement eventually.

Reply:
Good point!